snapcraft/storeapi/__init__.py

```diff
@@ -13,9 +13,285 @@
+            raise SnapNotFound(snap_name, channel, arch)
+        return self.download_snap(snap_name, channel, arch,
+                                  download_path, package['download_url'],
+                                  package['download_sha512'])
+
+    def get_macaroon(self, acl):
+        response = self.post('../../api/2.0/acl/{}/'.format(acl),
+                             data=json.dumps({}),
+                             headers={'Content-Type': 'application/json'})
+        if response.ok:
+            return response.json()['macaroon'], None
+        else:
+            return None, response.text
+
+    def get_discharges(self, email, password, one_time_password, macaroons):
+        data = dict(email=email, password=password,
+                    macaroons=[(k, v) for k, v in macaroons.items()])
+        if one_time_password:
+            data['otp'] = one_time_password
+        try:
+            response = self.sso.session.post(
+                '/tokens/discharge', data=data,
+                headers={'Content-Type': 'application/json',
+                         'Accept': 'application/json'})
+            return dict(response.content['discharge_macaroons']), None
+        except sso.ApiException as err:
+            return None, err.body
+        except sso.UnexpectedApiError as err:
+            return None, err.json_body
+
+    def upload_snap(self, name, data):
+        data['name'] = name
+        headers = {'Authorization': macaroon_auth(self.conf, 'package_upload')}
+        response = self.post('snap-upload/', data=data, headers=headers)
+        return response
+
+    def download_snap(self, name, channel, arch, download_path,
+                      download_url, expected_sha512):
+        if self.matching_sha512(download_path, expected_sha512):
+            logger.info('Already downloaded {} at {}'.format(
+                name, download_path))
+            return
+        logger.info('Downloading {}'.format(name, download_path))
+        download = self.cpi.get(download_url)
+        with open(download_path, 'wb') as f:
+            # FIXME: Cough, we may want to buffer here (and a progress bar
+            # would be nice) -- vila 2016-04-26
+            f.write(download.content)
+        if self.matching_sha512(download_path, expected_sha512):
+            logger.info('Successfully downloaded {} at {}'.format(
+                name, download_path))
+        else:
+            raise SHAMismatch(download_path, expected_sha512)
+
+    def matching_sha512(self, path, expected_sha512):
+        if not os.path.exists(path):
+            return False
+
+        file_sum = hashlib.sha512()
+        with open(path, 'rb') as f:
+            for file_chunk in iter(
+                    lambda: f.read(file_sum.block_size * 128), b''):
+                file_sum.update(file_chunk)
+        return expected_sha512 == file_sum.hexdigest()
+
+    # Low level helpers
+
+    def post(self, path, data, headers=None):
```

Comment:
I started to do so, and got confused half way: requests.Session define post and get, ssoclient.V2ApiClient defines post and get, every session wrapper defines post and.

Changing that rule just for the session wrappers defined here makes the code a constant puzzle: should I use get or _get ? _post ? Is that different from post ?

Otherwise, what's your definition of an internal helper as opposed to a "private" (even in the python lose sense) helper ?